Biomass_core.R

```diff
@@ -467,17 +467,33 @@ Init <- function(sim, verbose = getOption("LandR.verbose", TRUE)) {
   ##############################################
   ## species
   ##############################################
-  species <- setDT(sim$species)[, speciesCode := as.factor(species)]
-  LandR::assertColumns(species,
-                       c(species = "character", Area = "factor", longevity = "integer",
-                         sexualmature = "integer", shadetolerance = "numeric",
-                         firetolerance = "integer", seeddistance_eff = "integer",
-                         seeddistance_max = "integer", resproutprob = "numeric",
-                         resproutage_min = "integer", resproutage_max = "integer",
-                         postfireregen = "factor", leaflongevity = "integer",
-                         wooddecayrate = "numeric", mortalityshape = "integer",
-                         growthcurve = "numeric", leafLignin = "numeric",
-                         hardsoft = "factor", speciesCode = "factor"))
+  expectedClasses <- c(species = "character", Area = "factor", longevity = "integer",
+                       sexualmature = "integer", shadetolerance = "numeric",
+                       firetolerance = "integer", seeddistance_eff = "integer",
+                       seeddistance_max = "integer", resproutprob = "numeric",
+                       resproutage_min = "integer", resproutage_max = "integer",
+                       postfireregen = "factor", leaflongevity = "integer",
+                       wooddecayrate = "numeric", mortalityshape = "integer",
+                       growthcurve = "numeric", leafLignin = "numeric",
+                       hardsoft = "factor", speciesCode = "factor")
+  colsToFactor <- names(expectedClasses)[expectedClasses == "factor"]#c("speciesCode", "Area", "postfireregen", "hardsoft")
+  species <- setDT(sim$species)
+  set(species, NULL, "speciesCode", as.factor(species$species))
+  for (ctf in colsToFactor) {
+    set(species, NULL, ctf, as.factor(species[[ctf]]))
+  }
+  expectedColumns <- names(expectedClasses)
+  names(expectedColumns) <- names(expectedClasses)
+
+  allCorrect <- sapply(expectedColumns, function(n) {
+    is(species[[n]], expectedClasses[[n]])
+  })
+  if (sum(!allCorrect)) {
+    messageDF(data.frame(col = names(allCorrect), expectedClass = expectedClasses, correctClass = allCorrect )[!allCorrect])
+    stop("In sim$species object, some columns not correct type: see above" )
+  }
+  # LandR::assertColumns(species)
```

Comment:
why a custom assertion here? shouldn't we update the assertion itself, rather than implement a new one only here?

Reply:
Is this still necessary?

Reply:
I thought that the "rejected" code was the correct code here, no? @CeresBarros

Reply:
Yes, sorry github it seemed like you did the opposite. I checked the actual file now and it seems fine - couldn't test yet tho.